src/rust.rs

```diff
@@ -40,11 +53,82 @@ impl RustGenerator {
                 std::alloc::dealloc(ptr, layout);
             }
 
+            /// Converts a closure into a [`Waker`].
+            ///
+            /// The closure gets called every time the waker is woken.
+            pub fn waker_fn<F: Fn() + Send + Sync + 'static>(f: F) -> Waker {
+                let raw = Arc::into_raw(Arc::new(f)) as *const ();
+                let vtable = &Helper::<F>::VTABLE;
+                unsafe { Waker::from_raw(RawWaker::new(raw, vtable)) }
+            }
+
+            struct Helper<F>(F);
+
+            impl<F: Fn() + Send + Sync + 'static> Helper<F> {
+                const VTABLE: RawWakerVTable = RawWakerVTable::new(
+                    Self::clone_waker,
+                    Self::wake,
+                    Self::wake_by_ref,
+                    Self::drop_waker,
+                );
+
+                unsafe fn clone_waker(ptr: *const ()) -> RawWaker {
+                    let arc = ManuallyDrop::new(Arc::from_raw(ptr as *const F));
+                    core::mem::forget(arc.clone());
+                    RawWaker::new(ptr, &Self::VTABLE)
+                }
+
+                unsafe fn wake(ptr: *const ()) {
+                    let arc = Arc::from_raw(ptr as *const F);
+                    (arc)();
+                }
+
+                unsafe fn wake_by_ref(ptr: *const ()) {
+                    let arc = ManuallyDrop::new(Arc::from_raw(ptr as *const F));
+                    (arc)();
+                }
+
+                unsafe fn drop_waker(ptr: *const ()) {
+                    drop(Arc::from_raw(ptr as *const F));
+                }
+            }
+
+            #[cfg(target_family = "wasm")]
+            extern "C" {
+                fn __notifier_callback(idx: i32);
+            }
+
+            #[repr(transparent)]
+            pub struct FfiFuture<T: Send + 'static>(Pin<Box<dyn Future<Output = T> + Send + 'static>>);
+
+            impl<T: Send + 'static> FfiFuture<T> {
+                pub fn new(f: impl Future<Output = T> + Send + 'static) -> Self {
+                    Self(Box::pin(f))
+                }
+
+                pub fn poll(&mut self, _post_cobject: isize, port: i64) -> Option<T> {
+                    #[cfg(target_family = "wasm")]
+                    let waker = waker_fn(move || {
+                        unsafe { __notifier_callback(port as _) };
+                    });
+                    #[cfg(not(target_family = "wasm"))]
+                    let waker = waker_fn(move || unsafe {
+                        let post_cobject: extern "C" fn(i64, *const core::ffi::c_void) =
+                            core::mem::transmute(_post_cobject);
+                        let obj: i32 = 0;
+                        post_cobject(port, &obj as *const _ as *const _);
```

Comment:
What's `_post_cobject`?

Reply:
it's allows you to send a message to a dart `ReceivePort`. this is the only way to asynchronously notify dart (from an arbitrary thread)

Reply:
it's a pointer to a function in the dart c api